pkg/watermark/fetch/edge_fetcher_set.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2023 The Numaproj Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fetch
+
+import (
+	"context"
+	"math"
+	"time"
+
+	"github.com/numaproj/numaflow/pkg/isb"
+	"github.com/numaproj/numaflow/pkg/shared/logging"
+	"github.com/numaproj/numaflow/pkg/watermark/wmb"
+	"go.uber.org/zap"
+)
+
+// a set of EdgeFetchers, incoming to a Vertex
+// (In the case of a Join Vertex, there are multiple incoming Edges)
+// key=name of From Vertex
+type edgeFetcherSet struct {
+	edgeFetchers map[string]Fetcher
+	log          *zap.SugaredLogger
+}
+
+func NewEdgeFetcherSet(ctx context.Context, edgeFetchers map[string]Fetcher) Fetcher {
+	return &edgeFetcherSet{
+		edgeFetchers,
+		logging.FromContext(ctx),
+	}
+}
+
+// ProcessOffsetGetWatermark processes the offset on the partition indicated and returns the overall Watermark
+// from all Partitions
+func (efs *edgeFetcherSet) ProcessOffsetGetWatermark(inputOffset isb.Offset, fromPartitionIdx int32) wmb.Watermark {
+	_ = efs.ProcessOffset(inputOffset, fromPartitionIdx) // even if it errored, we'll keep going
+	return efs.GetWatermark()
+}
+
+// GetWatermark processes the Watermark for the given partition from the given offset
+func (efs *edgeFetcherSet) ProcessOffset(inputOffset isb.Offset, fromPartitionIdx int32) error {
+	var returnErr error
+	for _, fetcher := range efs.edgeFetchers {
+		err := fetcher.ProcessOffset(inputOffset, fromPartitionIdx)
+		if err != nil {
+			returnErr = err // instead of returning error immediately, first try processing offset for all edge fetchers
+		}
+	}
+	return returnErr
+}
+
+// GetHeadWatermark returns the latest watermark among all processors for the given partition.
+// This can be used in showing the watermark
+// progression for a vertex when not consuming the messages directly (eg. UX, tests)
+func (efs *edgeFetcherSet) GetHeadWatermark(fromPartitionIdx int32) wmb.Watermark {
+	// get the most conservative time (minimum watermark) across all Edges
+	var wm wmb.Watermark
+	overallWatermark := wmb.Watermark(time.UnixMilli(math.MaxInt64))
+	for fromVertex, fetcher := range efs.edgeFetchers {
+		wm = fetcher.GetHeadWatermark(fromPartitionIdx)
+		if wm == wmb.InitialWatermark { // unset
+			continue
+		}
+		efs.log.Debugf("Got Edge Head Watermark from vertex=%q while processing partition %d: %v", fromVertex, fromPartitionIdx, wm.UnixMilli())
+		if wm.BeforeWatermark(overallWatermark) {
+			overallWatermark = wm
+		}
+	}
+	return overallWatermark
+}
```

Comment:
**Rationale:**
EdgeFetcherSet is simply the addition of more "publishing processors", so any of the original logic that worked between Processors of a single EdgeFetcher can be extended between EdgeFetchers: in this case, find the minimum head watermark. 

Also, if `EdgeFetcher::GetHeadWatermark()` returns `wmb.InitialWatermark`, it means there was no Watermark found there. Just like `EdgeFetcher::GetHeadWatermark()` allows an individual Processor to not have a Watermark set, `EdgeFetcherSet::GetHeadWatermark()` allows for an entire EdgeFetcher not to have one set.